### ISSUE TITLE
Set adjoint `run_time` to forward source `fwidth` (if same `freq0`?)

### ISSUE THREAD

**Issue by tylerflex** [flexcompute/tidy3d#1403]
the current code for setting (single frequency) adjoint fwidth and run_time is below (paraphrased for simplicity)

```py
    @cached_property
    def _fwidth_adjoint(self) -> float:
        """Frequency width to use for adjoint source, user-defined or the average of the sources."""

        # if user-specified, use that
        if self.fwidth_adjoint is not None:
            return self.fwidth_adjoint

        freqs_adjoint = self.freqs_adjoint

        # multiple output frequency case
        if self._is_multi_freq:
            return FWIDTH_FACTOR_MULTIFREQ * self._min_delta_freq

        # otherwise, grab from sources and output monitors
        num_sources = len(self.sources)  # should be 0 for adjoint already but worth checking

        # if no sources, just use a constant factor times the mean adjoint frequency
        if num_sources == 0:
            return FWIDTH_FACTOR * np.mean(freqs_adjoint)

        # if more than one forward source, use their maximum
        if num_sources > 1:
            log.warning(f"{num_sources} sources, using their maximum 'fwidth' for adjoint source.")

        fwidths = [src.source_time.fwidth for src in self.sources]
        return np.max(fwidths)

```

```py
    @cached_property
    def _run_time_adjoint(self: float) -> float:
        """Return the run time of the adjoint simulation as a function of its fwidth."""

        if self.run_time_adjoint is not None:
            return self.run_time_adjoint

        return 100 / self._fwidth_adjoint
```

A problem is if the user's forward run_time is short for some reason (like 10/fwidth), then the adjoint one can be set to 100/fwidth, which was not ideal for a user who was running low on credits.

Perhaps to handle this we should add a new condition in the `_run_time_adjoint`

```py
# if both forward and adjoint are single frequency (with the same frequency) use forward simulation run_time.
if len(self.sources) == len(self.freqs_adjoint) == 1:
    if np.isclose(self.sources[0].source_time.freq0 == self.freqs_adjoint[0]):
        return self.run_time

```
? 
@momchil-flex  @tomflexcompute

**Comment by tylerflex** [comment]
Optionally, we could just reduce the `RUN_TIME_FACTOR=100` (`run_time_adj = RUN_TIME_FACTOR/fwidth`) to a more reasonable number between 10-100?

**Comment by momchil-flex** [comment]
I must say that having the `run_time_adj` depend on the `fwidth` maybe doesn't make too much sense. FDTD results are pretty much independent of fwidth as long as it covers the monitors frequencies.

I'm not sure what all the considerations you have in mind are. But roughly speaking, in which cases would it not make sense to just use the forward sim `run_time` in the adjoint sim too? Is it only because the adjoint source fwidth could be much smaller? In which case maybe we should just use something like `10 / self._fwidth_adjoint + self.run_time`? One could tinker on that a little if needed, e.g. reducing the `10` factor, or subtracting the forward source pulse time which should have been included in a sufficiently long `self.run_time`.

**Comment by tylerflex** [comment]
> I must say that having the run_time_adj depend on the fwidth maybe doesn't make too much sense. FDTD results are pretty much independent of fwidth as long as it covers the monitors frequencies.

Well the problem I think is not so much the fwidth dependence but simply that we need `run_time_adj` to be long enough to trigger early shutoff.  I think if we had something like #1267 it could help a bit here.

>  in which cases would it not make sense to just use the forward sim run_time in the adjoint sim too?
> Is it only because the adjoint source fwidth could be much smaller? 

This is the main consideration yea.

> In which case maybe we should just use something like 10 / self._fwidth_adjoint + self.run_time?
> One could tinker on that a little if needed, e.g. reducing the 10 factor, or subtracting the forward source pulse time which should have been included in a sufficiently long self.run_time.

I dont think this will work. Let's say the fwidth_adjoint is very small compared to the forward source. Then 10/fwidth_adjoint will be too short I think.  Also then the self.run_time is a bit irrelevant as it's set by the forward source fwidth.

Perhaps what we could consider is **rescaling** the `self.run_time` using the ratios of the `fwidth` of the adjoint and fwd sources.

```py
run_time_adj = self.run_time * (self.sources[0].fwidth / self._fwidth_adjoint)
```

In other words:
* If the adjoint fwidth is the same as the forward, use the forward run_time.
* If the adjoint fwidth is smaller than forward by a factor of `x`, we scale the run_time by `x`.

**Comment by momchil-flex** [comment]
> > I must say that having the run_time_adj depend on the fwidth maybe doesn't make too much sense. FDTD results are pretty much independent of fwidth as long as it covers the monitors frequencies.
> 
> Well the problem I think is not so much the fwidth dependence but simply that we need `run_time_adj` to be long enough to **reliably** trigger early shutoff, but not so long that it either misses the shutoff or creates headaches for users who want to estimate their costs. I think if we had something like #1267 it could help a bit here.

The problem is that I wouldn't consider this automatically reliable because of what we have already discussed e.g. even low-Q resonances are not automatically included, feedforward transmission is assumed in the automatic run time setting unless the user provides some target Q. But yeah if we had this it might not work directly but we *could* work on top of it. I think it might still involve taking some cues from the forward run_time and fwidth.

> > in which cases would it not make sense to just use the forward sim run_time in the adjoint sim too?
> > Is it only because the adjoint source fwidth could be much smaller?
> 
> This is the main consideration yea. There could be some additional considerations depending on where we place the adjoint vs. forward sources relative to the measurement locations but that seems like too much of an edge case to consider yet.
> 
> > In which case maybe we should just use something like 10 / self._fwidth_adjoint + self.run_time?
> > One could tinker on that a little if needed, e.g. reducing the 10 factor, or subtracting the forward source pulse time which should have been included in a sufficiently long self.run_time.
> 
> I dont think this will work. Let's say the fwidth_adjoint is very small compared to the forward source. Then 10/fwidth_adjoint will be too short I think. Also then the self.run_time is a bit irrelevant as it's set by the forward source fwidth.

Here's how I think about it. We need enough run time for the source pulse to completely decay, plus enough run time for the last injected fields to propagate out of the simulation. The `10 / fwidth_adjoint` addresses the first part of this. The factor can probably be decreased below 10. The `run_time` addresses the second part. The adjoint fields are propagating in the same simulation as the real fields. They are injected in a different location but probably undergo a lot of the same pathways, in reverse. So whatever `run_time` was sufficient for the last injected forward fields to decay, should also be sufficient for the last injected adjoint fields to decay (this is an overestimation by default since `run_time` includes both the forward source pulse decay and the field propagation time).

> Perhaps what we could consider is **rescaling** the `self.run_time` using the ratios of the `fwidth` of the adjoint and fwd sources.
> 
> ```python
> run_time_adj = self.run_time * (self.sources[0].fwidth / self._fwidth_adjoint)
> ```
> 
> In other words:
> 
> * If the adjoint fwidth is the same as the forward, use the forward run_time.
> * If the adjoint fwidth is smaller than forward by a factor of `x`, we scale the run_time by `x`.

I don't think that's right because of my explanation above.

**Comment by tylerflex** [comment]
> We need enough run time for the source pulse to completely decay, plus enough run time for the last injected fields to propagate out of the simulation. The 10 / fwidth_adjoint addresses the first part of this. The factor can probably be decreased below 10. The run_time addresses the second part. 

What I dont understand about this is that the forward `run_time` should already take both the source pulse width and field decay into account right? So it would seem to me that if the fwdith_adj == fwidth_fwd, we can just use forward `run_time`?  Unless like you said the adjoint fields are somehow propagating to other channels involving potentially high Q resonances but the forward fields are not. I think this should be a fair assumption and we could maybe provide some way to tweak things if not. Related to your comment:

> Whatever we do though we should also just have some user-exposed way to tweak this. So that in cases where it doesn't work well, like with this user, they can just manually fix.

We do offer the ability to set `JaxSimulation.adjoint_run_time` but maybe this isn't enough control? Again I feel like a `RunTimeSpec` would be the ideal.

Now I guess the question is how to take these considerations into account when handling the case where we have different `fwidth`s in forward and adjoint.

 I guess my equation is too simplistic as it assumes that the forward `run_time` is only set by the forward source `fwidth` and ignores the proportion of contribution of field decay on the forward run time.
```py
run_time_adj = self.run_time * (self.sources[0].fwidth / self._fwidth_adjoint)
```

If we follow your logic and assume that the fwd `run_time` is set by the addition of the source pulse length and the field decay (Q, propagation length, etc). Maybe we can express the forward source as something like (C~10)

```py
run_time_fwd = C / fwidth_fwd + FIELD_DECAY_TIME
```

and the adjoint source (with a different `fwidth`, but assuming the same `FIELD_DECAY_TIME` as a property intrinsic to the simulation) should be
```py
run_time_adj = C / fwidth_adj + FIELD_DECAY_TIME
```

So perhaps working backwards, if we know `fwidth_fwd` from the forward sources, we can solve for `FIELD_DECAY_TIME` and then set a `run_time_adj` as

```py
FIELD_DECAY_TIME = run_time_fwd - C / fwidth_fwd
run_time_adj = run_time_fwd - C / fwidth_fwd + C / fwidth_adj
```

All we would need to assume here I guess is a value for `C`. What do you think? Does it address your comments?

**Comment by momchil-flex** [comment]
That's basically exactly what I meant originally:


> In which case maybe we should just use something like `10 / self._fwidth_adjoint + self.run_time`? One could tinker on that a little if needed, e.g. reducing the 10 factor, or subtracting the forward source pulse time which should have been included in a sufficiently long self.run_time.

However now I feel like subtracting `C / fwidth_fwd` is over-engineering. It's OK if we're not as tight as possible on the adjoint run time estimate. We just need to have it better than just something like C / fwdith the way it is now, because this doesn't make sense for small vs. big, low-Q vs high-Q simulations.

**Comment by tylerflex** [comment]
Ok, I'll make a PR but maybe we should wait for 2.7 to release this so we have some time to test it a bit more without needing to do a last minute revert?

**Comment by momchil-flex** [comment]
Yeah, sure, it does not sound urgent, and any complaining user can be directed to manually set the adjoint fwidth.
